### PR TITLE
ICU-21174 Add "AndCheckError" versions of "emplaceBack()" and "create()"

### PR DESCRIPTION
**Pull request by hugovdm**
I intend to squash this and attribute the whole PR to @younies, with his approval - as I really just copied his code to move it one step up the class hierarchy.

##### Checklist

- [X] Issue filed: https://unicode-org.atlassian.net/browse/ICU-21174
- [X] Updated PR title and link in previous line to include Issue number
- [X] Issue accepted
- [ ] Tests included
- [ ] Documentation is changed or added

### CODE REVIEW

File: icu4c/source/common/cmemory.h [hunk 1]
```diff
@@ -771,6 +783,19 @@ class MaybeStackVector : protected MemoryPool<T, stackCapacity> {
         return this->create(args...);
     }
 
+    template <typename... Args>
+    T *emplaceBackAndCheckErrorCode(UErrorCode &status, Args &&... args) {
+        if (U_FAILURE(status)) {
```
Comment (sffc) [inline icu4c/source/common/cmemory.h:788]: Optional: Down here you can probably get away with

```
return this->createAndCheckErrorCode(status, args...);
```

Comment (hugovdm) [reply icu4c/source/common/cmemory.h:781]: Done. I intend to squash into the first commit, leaving it attributed to @younies

Comment (sffc) [reply icu4c/source/common/cmemory.h:781]: I think you changed it in the wrong place.

Is there a unit test dedicated to these classes? If there is, you should add a unit test.

Comment (hugovdm) [reply icu4c/source/common/cmemory.h:781]: Oh great. Partial attention bites again. :-( Sorry!

There are no direct tests for MemoryPool or MaybeStackVector yet. What I had failed to do was merge into the PR which uses it and run that.

Comment (hugovdm) [reply icu4c/source/common/cmemory.h:781]: I see the trie tests are also under test/intltest/ - so it seems the right location for tests for source/common/.

Comment (hugovdm) [reply icu4c/source/common/cmemory.h]: (I wonder why it is called "cmemory" anyway, when it's C++ code?)

**Comment by hugovdm** [comment]
- Tested with my local version of https://github.com/icu-units/icu/pull/52
- Squashed
- Half way through adding some unit tests, but I think there are good reasons for adding them as part of https://github.com/unicode-org/icu/pull/1220 instead of this PR.

So: post-squash, one more approve needed, then I can merge this, and update our development branches.

**Pull request merged**